google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+        if original_document:
+            original_queries = original_document.metadata.queries  # type: ignore
+
+        should_use_snaphot: bool = (
+            original_document is None
+            # equivalent to:
+            #   `if snapshot.read_time > original_document.snapshot.read_time`
+            or _helpers.compare_timestamps(
+                snapshot.read_time, original_document.snapshot.read_time,
+            )
+            >= 0
+        )
+
+        if should_use_snaphot:
+            self.documents[full_document_path] = _BundledDocument(
+                snapshot=snapshot,
+                metadata=BundledDocumentMetadata(
+                    name=full_document_path,
+                    read_time=snapshot.read_time,
+                    exists=snapshot.exists,
+                    queries=original_queries,
+                ),
+            )
+
+        if query_name:
+            bundled_document = self.documents.get(full_document_path)
+            bundled_document.metadata.queries.append(query_name)  # type: ignore
+
+        self._update_last_read_time(snapshot.read_time)
+        # Flush this if it was cached
+        self._deserialized_metadata = None
+        return self
+
+    def add_named_query(self, name: str, query: BaseQuery) -> "FirestoreBundle":
+        """Adds a query to the bundle, referenced by the provided name.
+
+        Args:
+            name (str): The name by which the provided query should be referenced.
+            query (Query): Query of documents to be fully loaded and stored in
+                the bundle for future access.
+
+        Example:
+
+            from google.cloud import firestore
+
+            db = firestore.Client()
+            collection_ref = db.collection(u'users')
+
+            bundle = firestore.FirestoreBundle('my bundle')
+            bundle.add_named_query('all the users', collection_ref._query())
+
+        Returns:
+            FirestoreBundle: self
+
+        Raises:
+            ValueError: If anything other than a BaseQuery (e.g., a Collection)
+                is supplied. If you have a Collection, call its `_query()`
+                method to get what this method expects.
+            ValueError: If the supplied name has already been added.
+        """
+        if not isinstance(query, BaseQuery):
+            raise ValueError(
+                "Attempted to add named query of type: "
+                f"{type(query).__name__}. Expected BaseQuery.",
+            )
+
+        if self.named_queries.get(name):
+            raise ValueError(f"Query name conflict: {name} has already been added.")
+
+        # Execute the query and save each resulting document
+        _read_time = self._save_documents_from_query(name, query)
+
+        # Actually save the query to our local object cache
+        self._save_named_query(name, query, _read_time)
+        # Flush this if it was cached
+        self._deserialized_metadata = None
```

Comment:
these two lines show up often, but setting `_deserialized_metadata` doesn't map to 'flush from cache' in my head. Can you explain this? Perhaps this might be worth putting as a `_flush_cache` method (silly for one call but might allow for a good place to expand on how this works)

Reply:
Moved to dedicated method with comment.